### PR TITLE
Adds Histogram Aggregator

### LINKED ISSUE #317: Implement Histogram aggregator
The metics export pipeline has support for configurable aggregators. In the SDK specification (WIP), https://github.com/open-telemetry/opentelemetry-specification/pull/347, four aggregators are specified that SHOULD be provided to support measure metric instruments.  One is currently missing from this set, the Histogram aggregator.  This aggregator should (probably) support both exponential bucket sizes and linear bucket sizes.

### PR DESCRIPTION
**Pull request by paivagustavo**
This adds a new Aggregator implementation `Histogram`. 

Fixes #317

It returns a `Buckets` which consists of the bucket's boundaries and the counts for each bucket. These counts are *not* cumulative, the reason for this is to make the update simpler and faster. Exporters *may* accumulate them if needed.

### CODE REVIEW

File: sdk/export/metric/aggregator/aggregator.go [hunk 1]
```diff
@@ -63,6 +63,20 @@ type (
 		Points() ([]core.Number, error)
 	}
 
+	// Buckets represents histogram buckets boundaries and counts.
+	//
+	// For a Histogram with N defined boundaries, e.g, [x, y, z].
+	// There are N+1 counts: [-inf, x], (x, y], (y, z], (z, +inf]
```
Comment (rghetia) [inline sdk/export/metric/aggregator/aggregator.go]: [opencensus metrics](https://github.com/census-instrumentation/opencensus-proto/blob/be218fb6bd674af7519b1850cdf8410d8cbd48e8/src/opencensus/proto/metrics/v1/metrics.proto#L209-L219) has [-inf, x), [x, y), [y, z), [z, +inf). It is a minor difference but could create compatibility issue.

Comment (paivagustavo) [reply sdk/export/metric/aggregator/aggregator.go]: That is interesting. I wonder what are the most used operator on histograms, `less than` or `less than or equals`. Prometheus, for example, uses the `less than or equal` operator.  The differences should be little in the end.

I'll leaning to change it to be `less than` to be consistent with OpenCensus, if anyone have another opinion here would be great.

File: sdk/metric/aggregator/histogram/histogram.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+func (c *Aggregator) Checkpoint(ctx context.Context, desc *export.Descriptor) {
+	// N.B. There is no atomic operation that can update all three
+	// values at once without a memory allocation.
+	//
+	// This aggregator is intended to trade this correctness for
+	// speed.
+	//
+	// Therefore, atomically swap fields independently, knowing
+	// that individually the three parts of this aggregation could
+	// be spread across multiple collections in rare cases.
```
[the remark this answers is not in the document]
Comment (paivagustavo) [reply sdk/metric/aggregator/histogram/histogram.go]: My plans are to merge this and then open an issue to track the proposed algorithm, what do you think?

[the remark this answers is not in the document]
Comment (rghetia) [reply sdk/metric/aggregator/histogram/histogram.go]: SG!

File: sdk/metric/aggregator/histogram/histogram.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+var _ aggregator.Count = &Aggregator{}
+var _ aggregator.Histogram = &Aggregator{}
+
+// New returns a new measure aggregator for computing Histograms.
+//
+// A Histogram observe events and counts them in pre-defined buckets.
+// And also provides the total sum and count of all observations.
+//
+// Boundaries MUST be ordered otherwise the histogram could not
+// be properly computed.
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/metric/aggregator/histogram/histogram.go]: Could we instead sort the boundaries? Panicking here seems like a bad user experience, especially when it is an error we can handle (by sorting).

File: sdk/export/metric/aggregator/aggregator.go (hunk 1 above)
Comment (MrAlias) [reply sdk/export/metric/aggregator/aggregator.go]: It looks like [the decision in open census](https://github.com/census-instrumentation/opencensus-proto/pull/121/commits/6543b80d05e3b2a65c93fbe3eae6679c7451c666#diff-ae7519afabde512d65fb5cb5ff33caa4R210-R220) was motivated by trying to remain compatible with [OpenMetrics](https://github.com/OpenObservability/OpenMetrics/pull/126/files#diff-d1792782ae179ab49b581de250c5def2R195-R203). That being said, it looks like it is still [an open issue](https://github.com/OpenObservability/OpenMetrics/issues/62) in the OpenMetric project.

I still think going with `(-inf, x), [x, y), [y, z), [z, +inf)` is the correct approach, and at this point I don't think it locks us into too much.

Comment (paivagustavo) [reply sdk/export/metric/aggregator/aggregator.go]: Thanks everyone for discussing this. I've changed the code to be consistent with the Open Census approach.

File: sdk/metric/aggregator/histogram/histogram_test.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+			if rand.Uint32() > math.MaxUint32/2 {
+				return -1
+			}
+			return 1
+		},
+	}
+
+	boundaries = []core.Number{core.NewFloat64Number(250), core.NewFloat64Number(500), core.NewFloat64Number(700)}
+)
+
```
[the remark this answers is not in the document]
Comment (paivagustavo) [reply sdk/metric/aggregator/histogram/histogram_test.go]: Done.

File: sdk/metric/aggregator/histogram/histogram.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	"go.opentelemetry.io/otel/api/core"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+	"go.opentelemetry.io/otel/sdk/export/metric/aggregator"
+)
+
+type (
+	// Aggregator observe events and counts them in pre-determined buckets.
+	// It also calculates the sum and count of all events.
+	Aggregator struct {
+		current    state
```
[the remark this answers is not in the document]
Comment (paivagustavo) [reply sdk/metric/aggregator/histogram/histogram.go]: Added this comment.

File: sdk/metric/aggregator/histogram/histogram.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		checkpoint state
+		boundaries []core.Number
+		kind       core.NumberKind
+	}
+
+	// state represents the state of a histogram, consisting of
+	// the sum and counts for all observed values and
+	// the less than equal bucket count for the pre-determined boundaries.
+	state struct {
+		Buckets aggregator.Buckets
```
[the remark this answers is not in the document]
Comment (paivagustavo) [reply sdk/metric/aggregator/histogram/histogram.go]: Added, thanks.

File: sdk/metric/aggregator/histogram/histogram.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (paivagustavo) [reply sdk/metric/aggregator/histogram/histogram.go]: Indeed that is a better experience. I've copied the boundaries slice and sorted it.

**Comment by paivagustavo** [comment]
I've addressed all the discussions, thanks everyone :)

File: sdk/metric/aggregator/histogram/histogram_test.go [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+		},
+		{
+			Name:   "state.count",
+			Offset: unsafe.Offsetof(state{}.count),
+		},
+	}
+	fmt.Println(fields)
+
+	if !ottest.Aligned8Byte(fields, os.Stderr) {
+		fmt.Println("QUEBROU")
```
Comment (MrAlias) [inline sdk/metric/aggregator/histogram/histogram_test.go]: ¡Ay, caramba!

Comment (paivagustavo) [reply sdk/metric/aggregator/histogram/histogram_test.go]: Caught in the act :see_no_evil:! Thanks for catching that, hahah.

**Pull request merged**